rfd/0127-resource-versioning.md

```diff
@@ -0,0 +1,34 @@
+
+## Details
+
+A Teleport resource contains, at a minimum, a `kind` defining what the resource is (like `role` or `web_session` or `kube_server`) and a `version`, which is a string that's currently always one of `v1`, `v2`, `v3`, `v4`, `v5`, `v6` (but is actually treated as a completely opaque identifier) and that is used to allow for changes in format or semantics across different versions of Teleport, to deliberately prevent the misinterpretation of a resource that might end up being accidentally compatible in terms of data. This is particularly relevant for resources such as `role`, whose semantics MUST be fully understood for a cluster component to be able to make access control decisions, lest they accidentally grant more access than intended.
+
+The flexibility of JSON as a storage format and ProtoBuf as an exchange format (as part of gRPC) lets us reuse essentially the same code to support different versions at once in almost all cases; this RFD proposes that we encode that as part of the `version` string of the resource, as such:
+
+- if fields are added, discarded or changed in such a way that older versions of Teleport will still behave "correctly enough" when interpreting the new resource data as the old one, the version doesn't need to change. This can and should be done deliberately whenever possible; for instance, a new field with broader semantics can be replicated onto an older, deprecated field, so that older versions of Teleport can still understand and use the data - see [#27018](https://github.com/gravitational/teleport/pull/27018) for an example of this in action.
+
+- if the format of the data stays compatible (in terms of JSON and ProtoBuf unmarshaling) but the meaning of the data changes enough that we want older versions of teleport to refuse parsing it, we bump the "minor" component of the version string; if the latest version of `role` is `v6`, such a change would introduce a `role` with version `v6.1`. This is not intended to be [Semantic Versioning](https://semver.org/), as there shouldn't be any need for interoperability outside of Teleport in the first place, so we are going to have a list of all the supported versions anyway. It's advisable to document which fields became available in which resource version, and to have checks in place so that resources that are marked to have a version don't actually end up specifying behavior or options that only became available in later versions - for instance, it would be advisable to reject any `v6` `role` that sets a field that was only introduced in `v6.1` (as that same role resource would be interpreted incorrectly by an older Teleport agent that only understood up to `v6`).
```

Comment:
Semantics aside, how does bumping the version to, say, `version: v6.1` instead of `version: v7` change things in practice? It's still a version bump that code needs to handle, no?

Also, how does this play with Kubernetes Operator and Terraform Provider? Can you add separate section on those as well?

Reply:
Yeah, I don't see how this really changes anything.

Though I would be really happy to see things like this go away:

```
kind: RoleV6
version: v4
```

Reply:
> Though I would be really happy to see things like this go away:

+1

We should only ever need "kind: role". Maybe we can retcon that as well?

Reply:
The point is to explicitly decouple the storage format from the version, so that all `vN.M` of a resource are stored in version `N`. We don't need that if we accept that we're going to store and manage the seventh version of the role resource in a struct called `RoleV6`.

Reply:
It's just a bit weird, but not much weirder than having a `role` resource with version `v5` stored in a struct called `RoleV6`, so I'd be on board with that.